# PROJECT NAME
# UBER REQUEST DATA- Exploratory Data Analysis (using sql and excel charts)

# PROBLEM STATEMENT
##### Perform EDA on uber supply and demand using pandas on uber request data, which includes pickup point(Airport, City), status(Trip completed, cancelled, no cars available), request time and drop time. The eda is done using sql and chart visualization and dashboard is created using google sheets.




# SUMMARY
##### This exploration is primarily done to uncover the demand and supply of the rides have been completed, number of rides cancelled either by not availabilty of cars or cancelled by drivers, what time of the day the rides are being taken or what day the maximum rides are completed. The data consist of 6 rows and 6745 enteries. We further look upon the suggestions and decisions (if applied) to be made to overcome problems for better service.

#### Github link - https://github.com/Anoushka-Thakur/My-activities

# KNOW YOUR DATA

### INSTALLING AND IMPORTING LIBRARIES

##### I have install pandasql, allows you to query pandas DataFrames using SQL syntax.

In [ ]:
pip install pandasql

In [ ]:
import pandas as pd
from pandasql import sqldf

### DATA LOADING

In [ ]:
# data loading
print('\n--Data loading and reviewing head and checking missing values--')
uber_df= pd.read_csv("/content/drive/MyDrive/Uber Request Data (2).csv")
# table has firt five head and last five tail
display(uber_df)





--Data loading and reviewing head and checking missing values--


Request id Pickup point  Driver id             Status  \
0            619      Airport        1.0     Trip Completed   
1            867      Airport        1.0     Trip Completed   
2           1807         City        1.0     Trip Completed   
3           2532      Airport        1.0     Trip Completed   
4           3112         City        1.0     Trip Completed   
...          ...          ...        ...                ...   
6740        6745         City        NaN  No Cars Available   
6741        6752      Airport        NaN  No Cars Available   
6742        6751         City        NaN  No Cars Available   
6743        6754         City        NaN  No Cars Available   
6744        6753      Airport        NaN  No Cars Available   

        Request timestamp       Drop timestamp  
0         11/7/2016 11:51      11/7/2016 13:00  
1         11/7/2016 17:57      11/7/2016 18:47  
2          12/7/2016 9:17       12/7/2016 9:58  
3         12/7/2016 21:08      12/7/2016 22:03  
4     13-07-2016 08:33:16  13-07-2016 09:25:47  
...                   ...                  ...  
6740  15-07-2016 23:49:03                  NaN  
6741  15-07-2016 23:50:05                  NaN  
6742  15-07-2016 23:52:06                  NaN  
6743  15-07-2016 23:54:39                  NaN  
6744  15-07-2016 23:55:03                  NaN  

[6745 rows x 6 columns]

### DATASET FIRST REVIEW

In [ ]:
# REVIEW
print(uber_df.head())

   Request id Pickup point  Driver id          Status    Request timestamp  \
0         619      Airport        1.0  Trip Completed      11/7/2016 11:51   
1         867      Airport        1.0  Trip Completed      11/7/2016 17:57   
2        1807         City        1.0  Trip Completed       12/7/2016 9:17   
3        2532      Airport        1.0  Trip Completed      12/7/2016 21:08   
4        3112         City        1.0  Trip Completed  13-07-2016 08:33:16   

        Drop timestamp  
0      11/7/2016 13:00  
1      11/7/2016 18:47  
2       12/7/2016 9:58  
3      12/7/2016 22:03  
4  13-07-2016 09:25:47  


### DATASET INFORMATION

In [ ]:
# display info
print('\n--Info--')
print (uber_df.info())


--Info--
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6745 entries, 0 to 6744
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Request id         6745 non-null   int64  
 1   Pickup point       6745 non-null   object 
 2   Driver id          4095 non-null   float64
 3   Status             6745 non-null   object 
 4   Request timestamp  6745 non-null   object 
 5   Drop timestamp     2831 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 316.3+ KB
None


### DATA SET ROWS AND COLUMNS COUNT

In [ ]:
# COUNT ROWS AND COLUMN
print('\n--Rows and columns--')
print(uber_df.shape)


--Rows and columns--
(6745, 6)


### UNDERSTANDING VARIABLES

In [ ]:
# Perform descriptive Analysis
print('\n--Descriptive Analysis')
print(uber_df.describe())



--Descriptive Analysis
        Request id    Driver id
count  6745.000000  4095.000000
mean   3384.644922   149.501343
std    1955.099667    86.051994
min       1.000000     1.000000
25%    1691.000000    75.000000
50%    3387.000000   149.000000
75%    5080.000000   224.000000
max    6766.000000   300.000000


###  DATA PREPROCESSING, CHECKING MISSING VALUES, EXTRACTING INSIGHTS

##### This columns is where data manipulation is done I have extracted request hour from request timestamp, days of the week from request dates, day of the time(morning, noon, evening, night by using def function)

##### In the data i have not drop null values as i have analysis the fulfilled trips and unfulfilled trips.

In [ ]:
# --- Data Cleaning/Preprocessing ---
# In the data i have not drop null values as i have analysis the fulfilled trips and unfulfilled trips


print('\n--Missing values--')
print(uber_df.isnull().sum())


# Convert 'Request timestamp' and 'Drop timestamp' to datetime
# Using errors='coerce' will turn unparseable dates into NaT (Not a Time)
uber_df['Request timestamp'] = pd.to_datetime(uber_df['Request timestamp'], errors='coerce')
uber_df['Drop timestamp'] = pd.to_datetime(uber_df['Drop timestamp'], errors='coerce')



# Extract 'Request Hour'
uber_df['Request Hour'] = uber_df['Request timestamp'].dt.hour

# Extract 'Request Day of Week' (Monday=0, Sunday=6)
uber_df['Request Day of Week'] = uber_df['Request timestamp'].dt.dayofweek



#extract day of the time (Morning, noon, evening, night, midnight)
# Extract day of the time

def get_time_slot(hour):
    if 0 <= hour <= 4:
        return 'Midnight'
    elif 5 <= hour <= 9:
        return 'Morning'
    elif 10 <= hour <= 14:
        return 'Noon'
    elif 15 <= hour <= 19:
        return 'Evening'
    else:
        return 'Night'
uber_df['Time slot']= uber_df['Request Hour'].apply(get_time_slot)




#strip string
for col in uber_df.select_dtypes(include='object').columns:
   uber_df[col] = uber_df[col].astype(str).str.strip()


--Missing values--
Request id              0
Pickup point            0
Driver id            2650
Status                  0
Request timestamp       0
Drop timestamp       3914
dtype: int64


# EDA using SQL (via pandasql)


In [ ]:
# Setup SQL function
pysqldf = lambda q: sqldf(q, globals())




##### Here is where our analysis start in sql, before i have used pandasql which is pandas library to execute sql query directly on pandas dataframes, its a great use for data manipulation and analysis in python environment.
##### The pysqldf is the variable for the function, instead of def I have use lambda, q is for argument, representing sql query, sqldf is the main function from pandasql library, globals(), it gains access to all the variables (including your pandas DataFrames).


#### 1. Request status distribution

##### First the analysi is of request status distribution. Where the count of the status like trip completed, cancelled and no cars available counts shows, trip completed has the highest count of 2831.

In [ ]:
# Request status distribution
q1= """
SELECT status, COUNT(*) AS count
FROM uber_df
GROUP BY status
"""
print(pysqldf(q1))

              Status  count
0          Cancelled   1264
1  No Cars Available   2650
2     Trip Completed   2831


#### 2. Request by pickup point

##### Second insight is on pickup point, the counts on airport has 3238 and city 3507.

In [ ]:
# Request by pickup point
q2= """
SELECT "Pickup point", COUNT(*) AS count
FROM uber_df
GROUP BY "Pickup point"
"""
print(pysqldf(q2))

  Pickup point  count
0      Airport   3238
1         City   3507


#### 3. Status by pickup point distribution

##### Status by pickup point thorws light on which point we have what status for instance airport has the most number of unavailabilty of car which 1713 then city.

In [ ]:
# Request status by pickup point
q3= """
SELECT "Pickup point", status, COUNT(*) AS count
FROM uber_df
GROUP BY "Pickup point", status
ORDER BY "Pickup point", COUNT (*) DESC
"""
print(pysqldf(q3))

  Pickup point             Status  count
0      Airport  No Cars Available   1713
1      Airport     Trip Completed   1327
2      Airport          Cancelled    198
3         City     Trip Completed   1504
4         City          Cancelled   1066
5         City  No Cars Available    937


#### 4. Status by time slot

##### Status by timeslot shows the day of time, for instance trip completed in  Morning has 360 , in Night 1885 Evening in 278 and Noon 231.

In [ ]:
# Request status by time slot
q4= """
SELECT "Time slot", status, COUNT(*) AS count
FROM uber_df
GROUP BY "Time slot", status
ORDER BY "Time slot", COUNT (*) DESC
"""
print(pysqldf(q4))

   Time slot             Status  count
0    Evening  No Cars Available    373
1    Evening     Trip Completed    278
2    Evening          Cancelled     53
3   Midnight  No Cars Available    105
4   Midnight     Trip Completed     77
5   Midnight          Cancelled     21
6    Morning     Trip Completed    360
7    Morning          Cancelled    323
8    Morning  No Cars Available    158
9      Night  No Cars Available   1932
10     Night     Trip Completed   1885
11     Night          Cancelled    809
12      Noon     Trip Completed    231
13      Noon  No Cars Available     82
14      Noon          Cancelled     58


#### 5. Hourly trend

##### Request hourly request query shows the hourly trip count made.The most trip were made between 18:00 to 20:00.

In [ ]:
# Request hourly trend
q5= """
SELECT "Request Hour", COUNT(*) AS count
FROM uber_df
GROUP BY "Request Hour"
ORDER BY "Request Hour"
"""
print(pysqldf(q5))

    Request Hour  count
0            NaN   4071
1            0.0     32
2            1.0     28
3            2.0     30
4            3.0     31
5            4.0     82
6            5.0    171
7            6.0    171
8            7.0    150
9            8.0    159
10           9.0    190
11          10.0     92
12          11.0     81
13          12.0     81
14          13.0     59
15          14.0     58
16          15.0     67
17          16.0     71
18          17.0    168
19          18.0    205
20          19.0    193
21          20.0    195
22          21.0    186
23          22.0    107
24          23.0     67


#### 6. Unfullfilled trips

##### This query show insight about unfulfilled trip or cancelled or no cars available, with count 2650.

In [ ]:
# Unfulfilled trips
q6= """
SELECT COUNT(*) AS count
FROM uber_df
WHERE status = 'No Cars Available'
"""
print(pysqldf(q6))


   count
0   2650


#### 7. Completion rate

###### This extra insight is done to see the percentage of completion rate means the total trip completed in percentage and by 6745 request made trip completed are 2831, in percentage its 41.97% which is not a good percent.

In [ ]:
# Completion rate
q7 = """
SELECT
  COUNT(*) as Total_Requests,
  SUM(CASE WHEN Status = 'Trip Completed' THEN 1 ELSE 0 END) as Completed,
  ROUND(SUM(CASE WHEN Status = 'Trip Completed' THEN 1.0 ELSE 0 END) / COUNT(*) * 100, 2) as Completion_Rate_Percent
FROM uber_df
"""
print(pysqldf(q7))


   Total_Requests  Completed  Completion_Rate_Percent
0            6745       2831                    41.97


# DATA VISUALIZATION AND UBER SUPPLY AND DEMAND INTERACTIVE DASHBOARD

####  GOOGLE SHEETS DASHBOARD LINK: https://docs.google.com/spreadsheets/d/13rnDM4Gvd-TVcaAFN5JURGm-ZohGO5B74-HmtodhQlc/edit?usp=sharing

# CONCLUSION

Based on the analysis, here are some conclusions, suggestions, and potential decisions:

**Conclusions:**

*   **High Number of Unfulfilled Requests:** A significant number of requests (cancellations and no cars available) indicate a clear imbalance between supply and demand.
*   **Time Slot Impact:** The "Night" time slot has the highest number of requests and also the highest number of unfulfilled requests, suggesting a major demand-supply gap during this period. The "Morning" time slot also shows a notable number of cancellations.
*   **Pickup Point Impact:** The Airport is a significant source of "No Cars Available" issues, while the City has a higher number of cancellations.
*   **Low Completion Rate:** The overall completion rate of around 42% is quite low, highlighting the severity of the supply-demand mismatch.

**Suggestions and Decision-Making:**

*   **Address the Night Time Slot Supply:** Focus on increasing the number of available drivers during the "Night" time slot, especially for Airport pickups. This could involve incentives for drivers working late hours or dynamic pricing during peak night hours from the Airport.
*   **Optimize Morning Peak in the City:** Investigate the reasons for cancellations in the "Morning" time slot from the City. This could be due to traffic, driver preferences, or inefficient matching. Strategies could include better route optimization, encouraging drivers to accept shorter trips during peak hours, or providing bonuses for completing trips in congested areas.
*   **Airport Strategy:** Develop a specific strategy for the Airport pickup point to reduce the "No Cars Available" issue. This might involve creating designated waiting areas for Uber drivers, improving communication between drivers and passengers at the Airport, or collaborating with Airport authorities to streamline the pickup process.
*   **Driver Incentives:** Implement targeted driver incentives based on time slots and pickup points to encourage drivers to be available in high-demand areas and during peak hours.
*   **Passenger Communication:** Provide more accurate wait time estimates to passengers, especially during peak hours and in areas with high demand. This can help manage passenger expectations and reduce cancellations due to long waits.
*   **Further Analysis:** Conduct deeper analysis into the reasons for driver cancellations and passenger cancellations. This could involve surveys, driver feedback, and analyzing trip details to identify patterns and root causes.
*   **Dynamic Pricing Review:** Evaluate the current dynamic pricing model to ensure it effectively balances supply and demand without alienating passengers.
*   **Consider Pooling Options:** Explore the feasibility of implementing ride-sharing or pooling options, especially during peak hours and for popular routes like Airport pickups, to maximize the utilization of available cars.

By implementing these suggestions, Uber can work towards improving its completion rate, reducing unfulfilled requests, and ultimately providing a better experience for both drivers and passengers.